File: server/encender.js


const ModbusRTU = require("modbus-serial");
const client = new ModbusRTU();

// --- Essential Configuration ---
const TARGET_IP = "192.168.7.10"; // PLC IP address
const TARGET_PORT = 502;          // Modbus TCP port
const SLAVE_ID = 1;               // Slave ID
const TIMEOUT = 5000;             // Timeout in milliseconds (5 seconds)
const PULSE_WIDTH_MS = 1000;       // Duration of the "press" in milliseconds (e.g., 200ms) - ADJUST AS NEEDED

// --- Button Addresses (Base 0) ---
// Proworx 000040 (Encender) -> Modbus base 0 address 39
const START_BUTTON_ADDRESS = 39;
// Proworx 000042 (Cancelar Alarma) -> Modbus base 0 address 41
const CANCEL_ALARM_BUTTON_ADDRESS = 39;
// --- End Button Addresses ---

// Helper function for delays
const sleep = (ms) => new Promise(resolve => setTimeout(resolve, ms));

/**
 * Simulates a momentary button press by writing True, waiting, then writing False.
 * @param {ModbusRTU} modbusClient The connected modbus-serial client.
 * @param {number} address The Modbus coil address to pulse.
 * @param {number} pulseDurationMs Duration of the pulse in milliseconds.
 * @returns {Promise<boolean>} True on success (both writes successful), False otherwise.
 */
async function pulseCoil(modbusClient, address, pulseDurationMs = PULSE_WIDTH_MS) {
    if (!modbusClient || !modbusClient.isOpen) {
        console.error("ERROR: Pulse coil failed: Client not connected or not open.");
        return false;
    }

    console.log(`INFO: Pulsing coil ${address}: Setting to TRUE...`);
    try {
        // 1. Press the button (Write True)
        await modbusClient.writeCoil(address, true);
        console.log(`INFO: Coil ${address} set to TRUE successfully.`);

        // 2. Wait for the pulse duration
        await sleep(pulseDurationMs);

        // 3. Release the button (Write False)
        console.log(`INFO: Pulsing coil ${address}: Setting back to FALSE...`);
        await modbusClient.writeCoil(address, false);
        console.log(`INFO: Coil ${address} set back to FALSE successfully.`);

        console.log(`INFO: Pulse completed for coil ${address}.`);
        return true;

    } catch (error) {
        console.error(`ERROR: Exception during pulse_coil for address ${address}: ${error.message}`);
        if (error.err) { // Check for specific Modbus error code if provided by the library
             console.error(`ERROR: Modbus Error Code: ${error.err}`);
        }
        // Attempting to set back to FALSE might fail if the first write failed and connection broke
        // Or if the second write itself failed. Robustly try to ensure it's off if possible,
        // but report the overall operation as failed.
        try {
             console.log(`INFO: Attempting recovery: Setting coil ${address} to FALSE after error...`);
             await modbusClient.writeCoil(address, false);
             console.log(`INFO: Recovery attempt: Coil ${address} set to FALSE.`);
        } catch (recoveryError) {
             console.error(`ERROR: Failed to set coil ${address} back to FALSE during error recovery: ${recoveryError.message}`);
        }
        return false;
    }
}

// --- Main Execution ---
async function main() {
    // --- Select the button to press ---
    // const targetButtonAddress = START_BUTTON_ADDRESS;
    // const buttonName = "START";
    const targetButtonAddress = CANCEL_ALARM_BUTTON_ADDRESS; // <-- Example: Press Cancel Alarm
    const buttonName = "CANCEL ALARM";                      // <-- Example: Press Cancel Alarm
    //----------------------------------

    console.log("----------------------------------------");
    console.log(`Attempting to PULSE Button: ${buttonName} (Address: ${targetButtonAddress})`);
    console.log(`Pulse width: ${PULSE_WIDTH_MS} ms`);
    console.log("----------------------------------------");

    let success = false; // Keep track of the operation status

    try {
        // Connect to the PLC
        console.log(`INFO: Connecting to ${TARGET_IP}:${TARGET_PORT}...`);
        await client.connectTCP(TARGET_IP, { port: TARGET_PORT });
        client.setID(SLAVE_ID);
        client.setTimeout(TIMEOUT);
        console.log(`INFO: Successfully connected to Modbus device.`);

        // Pulse the selected button
        console.log(`\n[ACTION] Pulsing ${buttonName} button...`);
        success = await pulseCoil(client, targetButtonAddress, PULSE_WIDTH_MS);

        if (success) {
            console.log(`[SUCCESS] Pulse command sequence for ${buttonName} completed successfully.`);
            console.log(`[INFO] Observe the system/PLC for the expected action.`);
        } else {
            // Error messages are already logged within pulseCoil
            console.error(`[FAILURE] Pulse command sequence for ${buttonName} failed.`);
        }

    } catch (error) {
        // Catch connection errors or unexpected issues in the main flow
        console.error(`ERROR: Connection or main execution error: ${error.message}`);
        success = false; // Ensure success flag is false
    } finally {
        // Ensure the connection is closed
        if (client.isOpen) {
            client.close(() => {
                console.log("INFO: Modbus client connection closed.");
                console.log("\nScript finished.");
            });
        } else {
            console.log("INFO: Modbus client connection was not open or already closed.");
            console.log("\nScript finished.");
        }
    }
}

// Run the main function
main();

File: server/encender.py
#!/usr/bin/env python3

from pymodbus.client import ModbusTcpClient
import time
import logging

# Configure logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')

# --- Essential Configuration ---
TARGET_IP = "192.168.7.10"  # PLC IP address
TARGET_PORT = 502           # Modbus TCP port
SLAVE_ID = 1                # Slave ID
TIMEOUT = 5                 # Timeout in seconds
PULSE_WIDTH = 0.2           # Duration of the "press" in seconds (e.g., 200ms) - ADJUST AS NEEDED

# --- Button Addresses (Base 0) ---
# Proworx 000040 (Encender) -> Modbus base 0 address 39
START_BUTTON_ADDRESS = 39
# Proworx 000042 (Cancelar Alarma) -> Modbus base 0 address 41
CANCEL_ALARM_BUTTON_ADDRESS = 41
# --- End Button Addresses ---

# --- Keep Deflector Addresses if needed elsewhere, otherwise can be removed ---
# DEFLECTOR_AUTO_ADDRESS = 50
# DEFLECTOR_MANUAL_ADDRESS = 49
# ---

def get_modbus_client():
    """Create and return a connected Modbus client"""
    client = ModbusTcpClient(
        host=TARGET_IP,
        port=TARGET_PORT,
        timeout=TIMEOUT
    )
    try:
        if not client.connect():
            logging.error(f"Failed to connect to Modbus device at {TARGET_IP}:{TARGET_PORT}")
            return None
        logging.info(f"Successfully connected to Modbus device at {TARGET_IP}:{TARGET_PORT}")
        return client
    except Exception as e:
        logging.error(f"Exception during connection to {TARGET_IP}:{TARGET_PORT}: {str(e)}")
        return None

# --- NEW FUNCTION: Pulse a coil ---
def pulse_coil(client, address, pulse_duration=PULSE_WIDTH):
    """
    Simulates a momentary button press by writing True, waiting, then writing False.
    Returns True on success (both writes successful), False otherwise.
    """
    if not client or not client.is_socket_open():
        logging.error(f"Pulse coil failed: Client not connected.")
        return False

    logging.info(f"Pulsing coil {address}: Setting to TRUE...")
    try:
        # 1. Press the button (Write True)
        result_on = client.write_coil(address, True, slave=SLAVE_ID)
        if result_on.isError():
            logging.error(f"Failed to write TRUE to coil {address}: {result_on}")
            return False
        logging.info(f"Coil {address} set to TRUE successfully.")

        # 2. Wait for the pulse duration
        time.sleep(pulse_duration)

        # 3. Release the button (Write False)
        logging.info(f"Pulsing coil {address}: Setting back to FALSE...")
        result_off = client.write_coil(address, False, slave=SLAVE_ID)
        if result_off.isError():
            # Important: Log error, but the press might have still worked
            logging.error(f"Failed to write FALSE back to coil {address}: {result_off}")
            # Decide if this constitutes failure. Usually yes, as the state is left ON.
            return False
        logging.info(f"Coil {address} set back to FALSE successfully.")

        logging.info(f"Pulse completed for coil {address}.")
        return True

    except Exception as e:
        logging.error(f"Exception during pulse_coil for address {address}: {str(e)}")
        return False

# --- Functions like read_deflector_status and set_deflector_manual_mode are no longer directly used ---
# --- You can keep them if needed for other purposes, or remove them. ---
# --- For clarity, I will comment them out for this specific button-pressing task ---
#
# def read_deflector_status(client):
#     ... (previous code) ...
#
# def set_deflector_manual_mode():
#      ... (previous code) ...
#
# def read_final_status_standalone():
#      ... (previous code) ...

# --- Main Execution ---
if __name__ == "__main__":
    # Select the button to press
    target_button_address = START_BUTTON_ADDRESS # Change to CANCEL_ALARM_BUTTON_ADDRESS to press the other one
    button_name = "START" if target_button_address == START_BUTTON_ADDRESS else "CANCEL ALARM"

    print("-" * 30)
    print(f"Attempting to PULSE Button: {button_name} (Address: {target_button_address})")
    print(f"Pulse width: {PULSE_WIDTH} seconds")
    print("-" * 30)

    client = None # Define client outside try block for finally clause
    success = False
    try:
        # Get a client connection
        client = get_modbus_client()

        if client:
            # Pulse the selected button
            print(f"\n[ACTION] Pulsing {button_name} button...")
            success = pulse_coil(client, target_button_address, PULSE_WIDTH)

            if success:
                print(f"[SUCCESS] Pulse command sequence for {button_name} completed successfully.")
                print(f"[INFO] Observe the system/PLC for the expected action.")
            else:
                print(f"[FAILURE] Pulse command sequence for {button_name} failed.")
        else:
            print("[FAILURE] Could not establish Modbus connection.")

    except Exception as e:
        logging.error(f"An unexpected error occurred in main execution: {str(e)}")
        print(f"[ERROR] An unexpected error occurred: {str(e)}")
    finally:
        # Ensure client is closed
        if client and client.is_socket_open():
            client.close()
            logging.info("Modbus client closed (Main Execution)")

    # Note: Verification by reading status might not be useful here,
    # as the coil should be False after the pulse. Success is determined
    # by whether the PLC performed the action triggered by the pulse.

    print("\nScript finished.")